src/prognostic_equations/cloud_fraction.jl

```diff
@@ -96,67 +88,53 @@ function quad_loop(env_thermo_quad::SGSQuadrature, vars, thermo_params)
+
+    qt_hat(χ) = qt_hat(χ, quadrature_type)
+    h_hat(χ) = h_hat(χ, quadrature_type)
+    ν_c(χ) = ν_c(χ, quadrature_type)
+
+    get_ts(χ) =
+        thermo_state(thermo_params; p = p_c, θ = h_hat(χ), q_tot = qt_hat(χ))
+    # cloudy/dry categories for buoyancy in TKE
+    f_cf(χ, ts) =
+        TD.has_condensate(thermo_params, ts) ? typeof(χ)(1) : typeof(χ)(0)
+    f_q_tot_sat(χ, ts) =
+        TD.has_condensate(thermo_params, ts) ? qt_hat(χ) : typeof(χ)(0)
+    function f(χ)
+        ts = get_ts(χ)
+        return (; cf = f_cf(χ, ts), q_tot_sat = f_q_tot_sat(χ, ts))
     end
 
-    @inbounds for m_q in 1:quad_order
-        if quadrature_type isa LogNormalQuad
-            qt_hat = exp(ν_q + sqrt2 * s_q * χ[m_q])
-            ν_c = ν_θ + s2_θq / max(s_q, eps_q)^2 * (log(qt_hat) - ν_q)
-        elseif quadrature_type isa GaussianQuad
-            qt_hat = qt_mean + sqrt2 * σ_q * χ[m_q]
-            μ_c = θl_mean + sqrt2 * corr * σ_θ * χ[m_q]
-        end
+    return quad(f, env_thermo_quad).cf
+end
 
+import ClimaCore.RecursiveApply: rzero, ⊞, ⊠
+function quad(f, quad_type)
+    χ = quad_type.a
+    weights = quad_type.w
+    quad_order = quadrature_order(quad_type)
+    FT = eltype(χ)
+    # zero outer quadrature points
+    T = typeof(f(χ[1]))
+    outer_env = rzero(T)
+    @inbounds for m_q in 1:quad_order
         # zero inner quadrature points
-        inner_env .= 0
-
+        inner_env = rzero(T)
         for m_h in 1:quad_order
-            if quadrature_type isa LogNormalQuad
-                h_hat = exp(ν_c + sqrt2 * s_c * χ[m_h])
-            elseif quadrature_type isa GaussianQuad
-                h_hat = (μ_c) + sqrt2 * σ_c * χ[m_h]
-            end
-
-            # condensation
-            ts = thermo_state(thermo_params; p = p_c, θ = h_hat, q_tot = qt_hat)
-            q_liq_en = TD.liquid_specific_humidity(thermo_params, ts)
-            q_ice_en = TD.ice_specific_humidity(thermo_params, ts)
-            T = TD.air_temperature(thermo_params, ts)
-            # autoconversion and accretion
-
-            # environmental variables
-            inner_env[i_ql] += q_liq_en * weights[m_h] * sqpi_inv
-            inner_env[i_qi] += q_ice_en * weights[m_h] * sqpi_inv
-            inner_env[i_T] += T * weights[m_h] * sqpi_inv
-            # cloudy/dry categories for buoyancy in TKE
-            if TD.has_condensate(q_liq_en + q_ice_en)
-                inner_env[i_cf] += 1 * weights[m_h] * sqpi_inv
-                inner_env[i_qt_sat] += qt_hat * weights[m_h] * sqpi_inv
-                inner_env[i_T_sat] += T * weights[m_h] * sqpi_inv
-            else
-                inner_env[i_qt_unsat] += qt_hat * weights[m_h] * sqpi_inv
-                inner_env[i_T_unsat] += T * weights[m_h] * sqpi_inv
-            end
-        end
-
-        for idx in 1:env_len
-            outer_env[idx] += inner_env[idx] * weights[m_q] * sqpi_inv
+            inner_env = inner_env ⊞ f(χ[m_h]) ⊠ weights[m_h] ⊠ FT(1 / sqrt(π))
```

Comment:
Why do we need RecursiveApply?

Reply:
This will allow us to compute quadrature over arbitrary variables in one pass (e.g., a vector of scalars, a NamedTuple of scalars, structs etc.)